interpreter/mesh/src/main/scala/io/buoyant/interpreter/mesh/Client.scala

```diff
@@ -0,0 +1,305 @@
+    }
+  }
+
+  private[this] val releaseNop = () => Future.Unit
+
+  /**
+   * Constructs a Var that, when observed, calls `open()` to obtain a
+   * gRPC stream, publishing updates into the Var's state.
+   *
+   * When the stream returns an error, the Var may be updated (as
+   * according to `toT`) and no further updates will will be
+   * published.
+   */
+  private[this] def streamVar[S, T](
+    init: T,
+    open: () => Stream[S],
+    toT: Try[S] => Option[T]
+  ): Var[T] = Var.async[T](init) { state =>
+    val rsps = open()
+
+    // As we receive streamed messages, we are careful not to release
+    // them until the state has been updated to a new value. This is
+    // intended to (1) integrate tightly with flow control and more
+    // importantly (2) later integrate with netty's reference counted
+    // bueffers.
+    @volatile var closed = false
+    def loop(releasePrior: () => Future[Unit]): Future[Unit] =
+      if (closed) Future.Unit
+      else rsps.recv().transform {
+        case Throw(e) =>
+          toT(Throw(e)) match {
+            case None =>
+            case Some(t) =>
+              state() = t
+          }
+          releasePrior().before(Future.exception(e))
+
+        case Return(Stream.Releasable(s, release)) =>
+          toT(Return(s)) match {
+            case None =>
+              release().before(loop(releasePrior))
+            case Some(t) =>
+              state() = t
+              releasePrior().before(loop(release))
+          }
+      }
+
+    val f = loop(releaseNop)
+    Closable.make { _ =>
+      closed = true
+      f.raise(Failure("closed", Failure.Interrupted))
+      rsps.reset(GrpcStatus.Ok())
+      Future.Unit
+    }
+  }
+
+  /**
+   * Constructs an ACtivity that, when observed, calls `open()` to obtain a
+   * gRPC stream, publishing updates into the Activity's state.
+   *
+   * When the stream returns an error, the state is updated to
+   * Activity.Failed and no further updates will will be published.
+   */
+  private[this] def streamActivity[S, T](
+    open: () => Stream[S],
+    toT: S => Option[T]
+  ): Activity[T] = {
+    val toState: Try[S] => Option[Activity.State[T]] = {
+      case Throw(e) => Some(Activity.Failed(e))
+      case Return(s) => toT(s) match {
+        case None => None
+        case Some(t) => Some(Activity.Ok(t))
+      }
+    }
+    Activity(streamVar(Activity.Pending, open, toState))
+  }
+
+  private[this] val decodeDtab: mesh.DtabRsp => Option[Dtab] = {
+    case mesh.DtabRsp(Some(mesh.VersionedDtab(_, Some(pdtab)))) =>
+      Some(fromDtab(pdtab))
+    case _ => None
+  }
+
+  private[this] val _collectFromEndpoint: PartialFunction[mesh.Endpoint, Address] = {
+    case mesh.Endpoint(Some(_), Some(ipBuf), Some(port), pmeta) =>
+      val ipBytes = Buf.ByteArray.Owned.extract(ipBuf)
+      val ip = InetAddress.getByAddress(ipBytes)
+      val meta = Seq.empty[(String, Any)] ++
+        pmeta.flatMap(_.authority).map(Metadata.authority -> _) ++
```

Comment:
I'm still not thrilled about explicitly listing the metadata fields in the proto and am still in favor of an untyped String->String map.